### PR TITLE
feat(skeleton): Slice B — bone hierarchy editing (#556)

### LINKED ISSUE #556: Skel: Slice B — Bone hierarchy editing (reparent / detach / attach / split / connect)
Parent epic: #554
Depends on: Slice A (#555)

## Goal

Beyond create/remove, riggers need to reshape the bone graph: reparent a chain to a different parent, detach a chain to root, **attach** a chain to another entity's skeleton (the user explicitly asked about this), split a long bone in two, connect/disconnect bones from their parent's tail.

## Scope

- **Reparent**: drag a bone to a new parent in the outliner, or right-click → "Set parent…". Children of the reparented bone follow.
  - Transform mode: pick "keep world position" (default, recomputes local TRS) or "keep local position" (TRS unchanged, parent change only).
- **Detach**: lift a bone (and its children) to be a child of the skeleton root.
- **Attach to entity's skeleton**:
  - Source: selected bones in entity A.
  - Target: entity B (must have a skeleton; one is created if absent).
  - Copies the bones into B's skeleton, preserving names (with collision-resolution rules: rename source `Tail` → `Tail.001` if B already has `Tail`).
  - Copies the corresponding `VertexBoneAssignment`s **only if** the user also selects the geometry to transfer. By default, attach is rig-only — the source bones move to entity B but the source mesh stays bound to entity A. Useful for adding a "weapon socket" bone to a character.
- **Split bone**: insert a new bone halfway down an existing one (or at a custom position along the bone's axis). Original bone becomes the parent, new bone is the child. Weight redistribution: linearly interpolated based on the split position.
- **Connect / disconnect**: toggle whether a bone's head coincides with its parent's tail (Blender convention).
- **Mirror chain**: duplicate a selected chain across an axis with `_l`/`_r` naming convention (depends on Slice E's naming-detection plumbing — coordinate; if Slice E hasn't landed, ship "duplicate without rename").
- **Undo**: new commands `ReparentBoneCommand`, `AttachBoneToEntityCommand`, `SplitBoneCommand`, `ConnectBoneCommand`. Each preserves the previous skeleton state for full revert.
- **MeshImporterExporter** round-trip: hierarchy changes survive glTF / FBX export.

## Acceptance Criteria

- [ ] Reparent in "keep world" mode visually moves the bone to its new parent with the same world transform.
- [ ] Reparent in "keep local" mode preserves the local TRS exactly.
- [ ] Detach moves a chain to root without breaking weight bindings.
- [ ] Attach-to-entity copies a "socket" bone from character A onto character B with the bone visible in B's outliner and gizmo controllable.
- [ ] Split bone produces two bones along the original axis with smooth weight interpolation.
- [ ] Connect / disconnect toggles head-tail coincidence correctly.
- [ ] Hierarchy changes round-trip through glTF (and FBX where possible).
- [ ] Every operation is undoable.
- [ ] Sentry breadcrumbs `scene.skel.hier.*` per op.
- [ ] Headless-CI tests for reparent (both modes), attach-to-entity, and split.

## Effort

~10 days. Attach-to-entity is the trickiest; cross-entity skeleton manipulation has edge cases on every existing entity referencing the source skeleton.

### PR DESCRIPTION
**Pull request by fernandotonon**
## Summary
- Add bone hierarchy ops: reparent (keep-world/local), detach, split, connect/disconnect, and rig-only attach-to-entity, via `SkeletonEditor` snapshot rebuild + undo commands
- Inspector UI: searchable reparent dialog (shows current parent), split/attach dialogs, two-row Skeleton Tools buttons, themed floating context menu (picker + viewport right-click)
- Skeleton debug overlay: joint spheres sized by bone length + thin parent→child link rods (clearer than one octahedron per child)

Closes #556

## Test plan
- [ ] Select a skinned mesh, enable Skeleton overlay — joints/links look correct; multi-child parents show one joint + thin links
- [ ] Reparent with keep-world and keep-local; undo/redo
- [ ] Detach selected bone; verify it becomes a root
- [ ] Split bone at ~0.5; weights still sum; undo
- [ ] Connect/disconnect toggles head gap; reconnect works for sole child
- [ ] Attach bone to another entity (name collision → `.001`); source unchanged
- [ ] Right-click bone in viewport and in Inspector bone picker → floating menu actions work
- [ ] Reparent dialog: filter/search parents, scroll long lists, current parent label shown
- [ ] `./build_local/bin/UnitTests --gtest_filter="SkeletonEditor*"`

Made with [Cursor](https://cursor.com)

## Summary by CodeRabbit

* **New Features**
  * Added skeleton editing tools for reparenting, detaching, splitting, connecting/disconnecting, and attaching bones between entities.
  * Added searchable dialogs for reparenting, splitting, and attaching bones.
  * Added a right-click bone context menu for quick access to editing actions.
  * Added undo/redo support for the new skeleton operations.
* **Style**
  * Improved skeleton visualization with clearer joints, hierarchy links, and selected-bone highlighting.

### CODE REVIEW

File: src/SkeletonEditor.cpp [hunk 1]
```diff
@@ -678,6 +704,415 @@ void SkeletonEditor::refreshAfterEdit(const std::string& entityName, const QStri
...
+    Snapshot snap = captureSnapshot(entity);
+    const unsigned short oldHandle = bone->getHandle();
+    const QString childName = uniqueBoneName(skel.get(), boneName + QStringLiteral("_split"));
+    const unsigned short newHandle = nextBoneHandle(skel.get());
+
+    Snapshot::BoneData splitBd;
+    splitBd.name = childName.toStdString();
+    splitBd.handle = newHandle;
+    splitBd.parentName = boneStd;
+    splitBd.position = axis * (1.f - t);
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/SkeletonEditor.cpp]: Fixed in 7afdd533 — the split joint is now placed at `axis * t`, and the test asserts tip length is preserved for `t=0.25`.

File: src/SkeletonEditor.cpp [hunk 2]
```diff
@@ -678,6 +704,415 @@ void SkeletonEditor::refreshAfterEdit(const std::string& entityName, const QStri
...
+    // Original bone keeps its parent; shrink its contribution to the tip by
+    // moving former children under the new split bone and scaling their locals.
+    for (auto& bd : snap.bones) {
+        if (bd.parentName == boneStd) {
+            bd.parentName = splitBd.name;
+            // Children were expressed in original bone space at full length;
+            // after the split joint they sit in the new bone's space past t.
+            bd.position = bd.position - axis * t;
+            bd.initialPosition = bd.initialPosition - axis * t;
+        }
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/SkeletonEditor.cpp:878]: Fixed earlier in 7afdd533 (joint at `axis * t`) and covered by the updated split test.

File: src/SkeletonEditor.cpp [hunk 3]
```diff
@@ -678,6 +704,415 @@ void SkeletonEditor::refreshAfterEdit(const std::string& entityName, const QStri
...
+    // Collect unique bones to copy (requested + descendants), parent-before-child.
+    std::unordered_set<std::string> toCopy;
+    for (const QString& name : boneNames) {
+        if (!srcSkel->hasBone(name.toStdString())) {
+            result.error = QStringLiteral("Source bone not found: %1").arg(name);
+            return result;
+        }
+        auto desc = collectDescendants(srcSkel.get(), name.toStdString());
+        toCopy.insert(desc.begin(), desc.end());
+    }
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/SkeletonEditor.cpp]: Fixed in 6f050558 — source bones are validated before `ensureEntitySkeleton` / destination mutation.

File: src/SkeletonEditor.cpp [hunk 4]
```diff
@@ -678,6 +704,415 @@ void SkeletonEditor::refreshAfterEdit(const std::string& entityName, const QStri
...
+        if (firstNewName.isEmpty())
+            firstNewName = unique;
+
+        Snapshot::BoneData bd;
+        bd.name = unique.toStdString();
+        bd.handle = nextHandle++;
+        if (srcBone->getParent() && toCopy.count(srcBone->getParent()->getName())) {
+            bd.parentName = renameMap[srcBone->getParent()->getName()];
+        } else {
+            bd.parentName = dstRootParent;
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/SkeletonEditor.cpp]: Fixed in 6f050558 — copied bones are ordered parent-before-child (no longer relying on handle order / `operator[]`).

File: src/SkeletonEditor.cpp [hunk 5]
```diff
@@ -678,6 +704,415 @@ void SkeletonEditor::refreshAfterEdit(const std::string& entityName, const QStri
...
+
+    if (connected) {
+        target->position = tip;
+        target->initialPosition = tip;
+    } else {
+        // Keep bind tip in initialPosition; offset only the live pose so
+        // isBoneConnected can recover the tip for sole children.
+        target->initialPosition = tip;
+        target->position = tip + dir * kDisconnectGap;
+    }
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/SkeletonEditor.cpp:981]: Acknowledged / deferred for this slice. Slice B intentionally stores connect state via pose only (no new persistent flag). Encoding a durable sole-child disconnect without a flag is ambiguous; tracking as follow-up if needed.

File: src/SkeletonDebug.cpp [hunk 6]
```diff
@@ -156,22 +169,33 @@ SkeletonDebug::~SkeletonDebug()
...
+        mBoneEntities.push_back(link);
+
+        const Ogre::Vector3 dir = childPos / length;
+        tp->setOrientation(Ogre::Vector3::UNIT_Y.getRotationTo(dir));
+        // Link mesh is unit length along +Y; keep thickness constant so
+        // multi-child fans don't look like a bundle of full bones.
+        const float thick = std::max(mBoneSize * 0.35f, 0.008f);
+        tp->setScale(thick, length, thick);
+
+        tagBoneVisual(link, pBone->getName(), mEntity->getName());
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/SkeletonDebug.cpp:198]: Acknowledged / deferred. Links are bind-pose hierarchy edges today (same as the old octahedrons). Live per-frame retarget during animation is a reasonable follow-up but out of scope for hierarchy-edit Slice B.
